text/projects/Rafal_Kilar/Hunt-Szymanski.tex

```diff
@@ -0,0 +1,156 @@
+
+Rozważamy problem najdłuższego wspólnego podciągu. Mamy dane dwa słowa $A[1..m]$ i $B[1..n]$. Przedstawimy algorytm Hunta-Szymanskiego działający w czasie $O((r+m) log\ n + n)$ dla słów z $r$ dopasowaniami i $d$ dominującymi dopasowaniami. Następnie przedstawimy jego usprawnioną wersję działającą w czasie $O(m\ log\ n + d\ log(2mn/d) + n)$.
+
+\section*{Algorytm Hunta-Szymanskiego}
+
+Algorytm i omówienie na podstawie pracy \cite{hunt1977fast}.
+
+Definiujemy wartości progowe $T_{i,k}$ jako 
+$$T_{i,k} = \text{ najmniejsze $j$ takie że $A[i..i]$ i $B[1..j]$ posiadają podciąg długości $k$}$$
+Możemy udowodnić kilka prostych faktów o wartościach progowych:
+
+\textbf{Lemat 1.} \textit{Jeśli $T_{i,1}, \dots, T_{i, k}$ są zdefiniowane, to $T_{i,1} < \dots < T_{i,k}$.}
+
+\textit{Dowód.} Zauważmy, że $B[T_{i,k}]$ jest ostatnim znakiem pewnego wspólnego podciągu $A[1..i]$ i $B[1..T_{i,k}]$. W przeciwnym przypadku $T_{i,k}$ nie byłoby minimalne. Wiemy więc, że $A[1..i]$ i $B[1..T_{i,k}-1]$ mają wspólny podciąg długości $k-1$, zatem $T_{i,k-1} \le T_{i,k}-1$.
+
+Lemat ten przyda nam się w konstrukcji algorytmu.
+
+\textbf{Lemat 2.} $T_{i,k-1} < T_{i+1, k} \le T_{i,k}$
+
+\textit{Dowód.} Ponieważ $A[1..i]$ i $B[1..T_{i,k}]$ mają wspólny podciąg długości $k$, to mają je także $A[1..i+1]$ i $B[1..T_{i,l}]$, więc $T_{i+1, k} \le T_{i,k}$. $A[1..i+1]$ i $B[1..T_{i+1,k}]$ mają wspólny podciąg długości $k$. Usunięcie końcowych znaków z tych słów zmniejsza długość najdłuższego wspólnego podciągu o co najwyżej $1$, więc $A[1..i]$ i $B[1..T_{i+1,k}-1]$ mają wspólny podciąg długości $k-1$, czyli $T_{i,k-1} \le T_{i+1,k} - 1$.
+
+Używając tego lematu możemy wyznaczyć wartości $T_{i,k}$
+
+\textbf{Lemat 3.}
+
+$$T_{i+1,k} = \begin{cases}
+    \text{najmniejsze $j$ takie że $A[i+1] = B[j]$ i $T_{i, k-1} < j \le T_{i,k}$} \\
+    T_{i, k} \text{ jeśli takie $j$ nie istnieje}
+\end{cases}$$
+
+\textit{Dowód.} Rozważamy przypadek, że takie $j$ nie istnieje. Z minimalności $T_{i+1,k}$, każde maksymalne wspólny podciąg $A[1..i+1]$ i $B[1..T_{i+1,k}]$ kończy się w $B[T_{i+1,k}]$. Z lematu 2. i założenia o $j$, $B[T_{i+1,k}] \neq A[i+1]$, więc rozważane podciąg jest także podciągiem $A[1..i]$ i $B[1..T_{i+1,k}]$, więc $T_{i,k} \le T_{i+1,k}$ i z lematu 2. $T_{i,k} = T{i+1,k}$.
+
+Załóżmy teraz, że takie minimalne $j$ istnieje. $A[1..i+1]$ i $B[1..j]$ mają wspólny podciąg długości $k$ - wspólne podciąg długości $k-1$ $A[1..i]$ i $B[1..T_{i,k-1}]$ powiększone o $A[i+1]$ i $B[j]$, więc $T_{i+1,k} \le j$. 
+
+Załóżmy niewprost, że $T_{i+1,k} < j$. Z lematu 2. mamy $T_{i,k-1} < T_{i+1,k}$, więc ostatni znak wspólnego podciągu $A[1..i+1]$ i $B[1..T_{i+1,k}]$ długości $k$ nie jest równy $A[i+1]$. $A[1..i]$ i $B[1..T_{i+1,k}]$ muszą również mieć podciąg długości $k$, więc $T_{i,k} \le T_{i+1,k}$. Z lematu 2. mamy więc równość $T_{i, k} = T_{i+1,k}$. Z założenia niewprost i wymogu $T_{i+1,k} < j \le T_{i,k}$ mamy $T_{i,k} \neq T_{i+1,k}$, co daje nam sprzeczność i kończy dowód.

```

Comment:
"ostatni znak wspólnego podciągu $A[1..i+1]$ i $B[1..T_{i+1,k}]$ długości $k$ nie jest równy $A[i+1]$"
I don't get it, please elaborate.